dev-docs/RFCs/v7.x/data-loading-rfc.md

```diff
@@ -12,108 +12,116 @@ This RFC suggests a number of improvements to loading support in deck.gl, in par
+      // or
+      fetch: url => fetch(url, {headers: {'Company-Access-Token': 'Secret-Value'}})
+    }
   })
 ```
 
-Design notes:
-- This overload would support declarative usage (json/pydeck/...).
+Design Notes:
+- A problem is that some loaders (mainly image loaders) do not support a separate call to fetch, but instead load and parse in a single operation. Cross origin type options must be specified in some other way. This is discussed in the loaders.gl RFC
 
+### Proposal 2a: Specify loader options for a layer
 
-## Proposal: Add `parse` prop to specify how data should be parsed
+The proposed strategy is to invest in refining option handling in loaders.gl and focus on having a simple and clean mechanisms in deck.gl to forward loader props to loaders.gl `load` call.
 
 ```js
-import {parse, registerLoaders} from '@loaders.gl/core';
-import {CSVLoader} from '@loaders.gl/csv';
-registerLoaders(CSVLoader);
-
-new AnyLayer({
-  coordinateSystem: COORDINATE_SYSTEM.IDENTITY,
-  data: CSV_URL,
-  // NEW: Accept a parse methods
-  parse,
-});
+  SomeBitmapLayer({
+    data: DATA_URL,
+    bitmap: BITMAP_URL,
+    loadOptions: {
+      fetch: {headers: {...}}
+    }
+  })
 ```
 
-There are two ways to specify a custom loader that works with the deck.gl async prop loading:
+Design Issues:
+- Naming of the prop: `load`, `loadOptions`, `loaderOptions`, ...?
+- Do we need to be able to specify some loader options in the layer default props?
+- If so, do we need to implement a deep merge of loader prop objects? Should loaders.gl support that?
 
-- Defining a fetch function that loads and parses.
 
-Do we expect `parse` functions to support different input data.
+### Proposal 2b: Ability to override fetch per async prop
 
+Problem: different resources may be served from different servers and they may need different headers etc.
 
-The `parse` prop is expected to be either a function or a (list of) loaders.gl loaders that can be passed to `parse`.
-
-Design Notes:
-- The requirement to wrap custom loaders as loaders.gl loader objects: loaders.gl `parse` is flexible and accepts a fetch response object as its preferred (most  flexiable/efficient input). It is fair to assume that not all custom loaders support a variety of objects. By registering a custom loaders
-
-
-
-## Proposal 3a: Ability to override fetch per async prop
-
-By supplying an object to fetch, different request options could be needed for different props (different resources may be served from different servers and they may need different headers etc):
+By supplying an `props` object to the loader objects, different `load` options could be needed for different props.
 
 ```js
   SomeBitmapLayer({
     data: DATA_URL,
     bitmap: BITMAP_URL,
-    fetch: {
-      data: url => fetch(url, {headers: {...}})
-      bitmap: url => fetch(url, {headers: {...}})
+    loadOptions: {
+      props: {
+        data: {
+          fetch: url => companyInternalFetch(url)
```

Comment:
I personally dislike nested callbacks. The recursive type check can be exhausting. On the top level we could support both object and callback supplied to `loadOptions`:

```
loadOptions: (url, {propName, layer, ...}) => {...}
```

Reply:
> The recursive type check can be exhausting. 

The way I thought about it there would be no recursive typecheck.

The proposed `props` mechanism was intended to be a simple generic "override merge" system. I.e. before passing on the options, deck checks if there is a `props.loadOptions.props[propName]` object and merges that over the default `props.loadOptions` (and then deletes `loadOptions.props`). 

This way all the options handling is done by loaders.gl, and the override mechanism is completely generic (it could even be activated by a prop type flag, as deck doesn't know anything about the semantics of these props, so the same pattern could be reused for other deck.gl props).

The nested prop structure has the (to me) big advantage of supporting declarative overrides per prop, which extends immediately to json and pydeck. (I suppose this could be mitigated by adding support for expression evaluation for non-accessor props, but I was trying to avoid going there for now). 

In this proposal we don't pass the layer (and the prop name) into the fetch function but I think that is OK, keeps it clean. loaders.gl doesn't know about layers and props, and the designs where it has been used (at least ScenegraphLayer) need to be redone anyway to ensure that loaders can be called "idempotently" inside and outside of async props, that has been a major pain point for 3D tiles.

A possible discussion point with this proposal is that it could stand to be API audited next to simplar "prop map" type props like `updateTriggers` and `transitions` (and those could probably also stand to be re-audited).